File: ruby/base/action_handler.rb
require 'json'
require_relative './database'

class ActionHandler
  attr_reader :body, :headers, :db

  def initialize(body, headers)
    @body = JSON.parse(body.string, symbolize_names: true)
    @headers = headers
    log("ACTION: #{action} STARTED")
  end

  def initdb(namespace)
    return if testing

    @db = Database.new(
      host: secret('wov-db-host'),
      dbname: secret("#{namespace}-db-name"),
      user: secret("#{namespace}-db-user"),
      password: secret("#{namespace}-db-pw")
    )
  end

  def log(message)
    return if testing

    puts message
  end

  def testing
    ENV['RUBY_ENV'] == 'test'
  end

  def action
    body.dig(:action, :name)
  end

  def input(key)
    body.dig(:input, key)
  end

  def hasura(key)
    hasura_key = "x-hasura-#{key}".to_sym
    body.dig(:session_variables, hasura_key)
  end

  def response(payload)
    {
      body: payload.to_json,
      headers: {'content-type' => 'application/json'},
      code: 200
    }
  end

  def secret(name)
    path = "/var/openfaas/secrets/#{name}"
    return nil unless File.exist?(path)

    File.read(path).gsub("\n", '')
  end
end

File: template/python/base/action_handler.py
import os
from database import Database


class ActionHandler:
    """Common helper methods for Hasura Actions"""

    def __init__(self, event, context):
        self.body = event.body
        self.headers = event.headers
        self.namespace = os.environ.get('appname')
        self.secrets_path = (os.environ.get('secrets')
                             or '/var/openfaas/secrets')
        self.initialize_db()

    def initialize_db(self):
        if self.testing:
            return
        self.db = Database(
            host=self.secret('wov-db-host'),
            database=self.secret(f'{self.namespace}-db-name'),
            user=self.secret(f'{self.namespace}-db-user'),
            password=self.secret(f'{self.namespace}-db-pw')
        )

    def secret(self, name):
        path = f'{self.secrets_path}/{name}'
        with open(path, 'r') as file:
            return file.read()

    def log(self, message):
        if self.testing():
            return
        print(message)

    def testing(self):
        return os.environ.get('TESTING')

    def action(self):
        return self.body['action']['name']

    def input(self, key):
        return self.body['input'][key]

    def hasura(self, key):
        hasura_key = f'x-hasura-{key}'
        return self.body['session_variables'][hasura_key]

    def response(self, payload):
        return {
            'body': payload,
            'headers': {'content-type': 'application/json'},
            'code': 200
        }
